### ISSUE TITLE
Multiple GPU training is slow.

### ISSUE THREAD

**Issue by Ziba-li** [NVlabs/neuralangelo#57]
Dear author, Thanks for seeing this question.
When I trained toy_example with an eight-card 4090GPU server, I found that the training speed was not very fast. Similar to single card training. And it takes more than 100 hours to complete 500,000 epochs for single-card training. I understand that eight-card distributed training should take more than ten hours to complete, but the actual speed is not much faster. What is the reason for this?
![image](https://github.com/NVlabs/neuralangelo/assets/90883114/efe6836b-821b-4c4e-9848-446498bd9caf)
![image](https://github.com/NVlabs/neuralangelo/assets/90883114/467ab998-6723-4dc9-b0a3-71b0e9b9de65)

**Comment by xiemeilong** [comment]
I only need 24 hours to train 500,000 epochs with a single 4090 GPU. However, when I use 4x 4090 GPUs, the training speed is actually only a quarter of using 1 GPU. I have the same issue with unofficial implementations. I guess it's because the model size is so big that the communication overhead between GPUs for syncing gradients is too large.

**Comment by Ziba-li** [comment]
> I only need 24 hours to train 500,000 epochs with a single 4090 GPU. However, when I use 4x 4090 GPUs, the training speed is actually only a quarter of using 1 GPU. I have the same issue with unofficial implementations. I guess it's because the model size is so big that the communication overhead between GPUs for syncing gradients is too large.

Did you run the following script to get the data?
```
EXPERIMENT_NAME=toy_example
PATH_TO_VIDEO=toy_example.MOV
SKIP_FRAME_RATE=24
SCENE_TYPE=object #{outdoor,indoor,object}
bash projects/neuralangelo/scripts/preprocess.sh ${EXPERIMENT_NAME} ${PATH_TO_VIDEO} ${SKIP_FRAME_RATE} ${SCENE_TYPE}
```

I ran it according to the process, but I don't understand why it takes 2.3 seconds for an epoch to train on a 4090 GPU. 500,000 epochs will take 319 hours.

**Comment by xiemeilong** [comment]
I used my own data, only over 500 images.

**Comment by Ziba-li** [comment]
> I used my own data, only over 500 images.

The toy data I used actually only has 29 valid pictures. I don't understand why it takes 319 hours to complete 500,000 epochs.

**Comment by smandava98** [comment]
+1. How can I make training faster? It's impossible to run this in multiple videos as each takes so long. I'm looking for faster training for each video

**Comment by chenhsuanlin** [comment]
Hi @Ziba-li, the multi-GPU setup (i.e. distributed training) enables training with larger batch sizes. It doesn't increase the per-iteration training speed, but it will be much faster to train each epoch. If you want to look into training with less iterations, [this](https://github.com/NVlabs/neuralangelo/issues/38#issuecomment-1687633673) may be helpful.

**Comment by Ziba-li** [comment]
> Hi @Ziba-li, the multi-GPU setup (i.e. distributed training) enables training with larger batch sizes. It doesn't increase the per-iteration training speed, but it will be much faster to train each epoch. If you want to look into training with less iterations, [this](https://github.com/NVlabs/neuralangelo/issues/38#issuecomment-1687633673) may be helpful.

Thank you for your reply, but I still haven't understood why it takes a very long time for a single GPU to run for 500,000 epochs, instead of running 16 hours for the A100 24G to get the results.

**Comment by otakudj** [comment]
I have the same problem. When I use single 3090 gpu running the demo lego, it takes ~9secs an epoch. 
![image](https://github.com/NVlabs/neuralangelo/assets/27484315/8779c5fa-5ea8-421e-bc12-7eea3ccb9606)
Howerver, when I use 3090x4 to train the model, it takes ~23secs, which is very strange.
![image](https://github.com/NVlabs/neuralangelo/assets/27484315/1266a920-cd9c-47be-800e-283f7e3cd93f)

**Comment by prettybot** [comment]
I just tried with A100-PCIE-40GB on AutoDL platform.
With single GPU, the training time for lego.mp4 is as the following:
![img_v2_2509e97c-db3d-4205-b8bd-0b6e803aa97g](https://github.com/NVlabs/neuralangelo/assets/25277529/74c18ddb-838a-4174-aa6b-1f0d6be80a48)

For the hyperparameters part, I only update the dict_size from 22 to be 21.

**Comment by chenhsuanlin** [comment]
I'm not sure about the communication overhead of 4090, but we didn't see such issue with A100. If you could help pinpoint where the additional overhead is coming from (and verifying that it is indeed coming from the gradient synchronization part), I can put up a note on that.

Also a minor note that we are measuring by iterations (500k) instead of epochs in the codebase.

**Comment by prettybot** [comment]
@chenhsuanlin thanks a lot for your explaination about the 500k part.

**Comment by uu5208** [comment]
> I just tried with A100-PCIE-40GB on AutoDL platform. With single GPU, the training time for lego.mp4 is as the following: ![img_v2_2509e97c-db3d-4205-b8bd-0b6e803aa97g](https://user-images.githubusercontent.com/25277529/263332876-74c18ddb-838a-4174-aa6b-1f0d6be80a48.jpg)
> 
> For the hyperparameters part, I only update the dict_size from 22 to be 21.

Hello,  I trained the lego-demo with 2x A100-PCIE-80GB. But I really get the bad time consumption result as below. For each epoch I need 7s. 
![image](https://github.com/NVlabs/neuralangelo/assets/32358008/7d3d6cf8-e4fb-4f7d-944e-a7d1707a3d19)

**Comment by JLee2022** [comment]
I got the same problem.....The same as the problem in sdfstudio.
